crates/neon/src/instance/mod.rs

```diff
@@ -0,0 +1,72 @@
+use std::any::Any;
+use std::marker::PhantomData;
+use std::sync::atomic::{AtomicUsize, Ordering};
+
+use once_cell::sync::OnceCell;
+
+use crate::context::Context;
+use crate::lifecycle::InstanceData;
+
+static COUNTER: AtomicUsize = AtomicUsize::new(0);
+
+fn next_id() -> usize {
+    COUNTER.fetch_add(1, Ordering::SeqCst)
+}
+
+/// A cell that can be used to allocate data that is global to an instance
+/// of a Neon addon.
+pub struct Global<T> {
+    _type: PhantomData<T>,
+    id: OnceCell<usize>,
+}
+
+impl<T> Global<T> {
+    pub const fn new() -> Self {
+        Self {
+            _type: PhantomData,
+            id: OnceCell::new(),
+        }
+    }
+
+    fn id(&self) -> usize {
+        *self.id.get_or_init(next_id)
+    }
+}
+
+impl<T: Any + Send> Global<T> {
+    pub fn borrow<'a, 'b, C>(&self, cx: &'b mut C) -> Option<&'b T>
```

Comment:
Nit, even though we use it in other places, I get confused which is `'a` and which is `'b`. What do you think of trying to call the inner lifetime `'cx` more frequently?

```suggestion
    pub fn borrow<'a, 'cx, C>(&self, cx: &'a mut C) -> Option<&'a T>
    where
        C: Context<'cx>,
```